src/tasks/EmitBundleTask/EmitBundleBase.cs

```diff
@@ -0,0 +1,400 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+public abstract class EmitBundleBase : Microsoft.Build.Utilities.Task, ICancelableTask
```

Comment:
I think there will be an ask to be able to generate one C source and one object file that includes data + pre-alloc + register, or at least have an option to combine pre-alloc and register functions into the same source/object, and keep the raw data in separate source/object files. Any reason to keep the pre-alloc + registration code in separate source files to begin with? They are close related, since the registration uses the pre-alloc data, so maybe we should just have one source/object file for that data.

Reply:
Good point, I think the pre-alloc and registration code were separated because there was an idea to generate the registration code outside of the EmitBundle task (i.e. in the LibraryBuilder task), but when pivoting back to having the EmitBundle task generate the registration code, I overlooked being able to merge the two together.
The benefit to having the two separate would just to make the registration code more accessible, as the preallocated code can take several hundred lines.

Reply:
Since the registration code uses the preallocated structs there is already a dependency between them, and as you say, the registration code is far less then preallocated structs, so even if you want to use the preallocated structs only from that source file, the additional code for the registration should be minimal, and since you won't call the registration functions (if you do your own registration based on the preallocated structs), it will be linked out by linker in the end.

Reply:
What about the generated data symbols, they are generated into a separate source/object file, correct? Is it possible to include several bundle resource data into the same generated source/object file?

I could see the following use cases of EmtiBundleTask:

Call EmitBundle with one resource, then I would expect to get one source/object file including the data/len symbols + one source/object file with preallocated bundle + registration code.

Call EmitBundle with multiple resources, then I would expect to get one source/object file including data/len symbols for all resources + one source/object file with reallocated bundles + registration code for all resources.

Call EmitBundle with multiple resource, then I would expect to get one source/object file including data/len symbols for each resource + one source/object file with preallocated bundles + registration code for all resources.

For all uses cases above, have an option when calling EmitBundleTask to only emit the data/len symbols and skip generating source/object file including preallocated bundles + registration code.

EmitBundleTask should have an output item where it reports the name of all data/len symbols it has emitted. If we follow a pattern for the symbol names, [resource_name]_data and [resource_name]_len then it will be enough to report back all [resource_name] emitted, since caller of task could then build up its own use of emitted data and len symbols. This could be a use case if you just want to use EMitBundleTask to generate the embedded resources, but would like to create your own bundling logic or consume the generate symbols in a complete different way than using our existing bundling API's.

Reply:
Thanks for the suggestion,
Now the EmitBundle task does the following:

Always
- Generates 1 (1 for all) OR n (1 for each) source/object file(s) to contain the data/len symbols for the n resources

Optional if passed `BundleFile` parameter
- Generates 1 source/object file that preallocates `MonoBundled*Resource` structs and holds the function that registers the `MonoBundled*Resource` preallocated structs
- Generates 1 header file that is a combination of `bundled-resources-internals.h` and the `MonoBundledSatelliteAssembly` struct definition in `loader-internals.h`

Outputs (if the msbuild task asks for the output)
- An `ITaskItem[]` that is a copy of the resources being bundled (after filtering, grouping, and grabbing the first ITaskItem of each group) that appends the data and len symbols as `DataSymbol` `LenSymbol` metadata (naming up for suggestions)